ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Util/MonadSTM/StrictMVar.hs

```diff
@@ -168,7 +168,23 @@ isEmptyMVar StrictMVar { tmvar } = atomically $ Lazy.isEmptyTMVar tmvar
 
 updateMVar :: (MonadSTM m, HasCallStack) => StrictMVar m a -> (a -> (a, b)) -> m b
 updateMVar StrictMVar { tmvar, tvar, invariant } f = do
-    (a', b) <- atomically $ do
+    -- it's not unreasonable to assume that forcing !(!a', b) inside the
+    -- atomically block will force the new value before putting it into the
+    -- MVar, but although the value in the tuple is forced, there's actually
+    -- a thin closure constructed that just points to the forced value which
+    -- is what GHC returns in the constructed tuple (so it is actually a thunk,
+    -- albeit a trivial one!). in order to ensure that we're forcing the value
+    -- inside the MVar before calling checkInvariant, we need an additional
+    -- bang outside the atomically block, which will correctly force a' before
+    -- checkInvariant looks to see if it's been evaluated or not. without this
+    -- change, it's possible to put a lazy value inside a StrictMVar (though
+    -- it's unlikely to occur in production environments because this
+    -- intermediate unforced closure is optimized away at -O1 and above).
+    --
+    -- To exactly see what we mean, compile this module with `-ddump-stg-final`
+    -- and look for the definition of the closure that is placed as the first
+    -- item in the returned tuple here
```

Comment:
As we say "the returned tuple here" I would put this next to 
```
return (a', b)
```
instead of here. Sorry that I didn't say it on my previous comment

Reply:
ah! I read "the returned tuple here" to mean "the tuple returned from the `atomically` block", I'll change the comment slightly to "the tuple returned here" and put it before the last line of the `atomically`